### PR TITLE
The base class for SCIM provisioning

### PR DESCRIPTION
**Pull request by shavantha**
Purpose
This PR contains the base class for SCIM provisioning. The test cases that add delete users will be calling this base class.

1.ProcessJsonFile.java : This class will read the json paylpoad and create the json object that will contain the data for provisioning

2.SCIM2TestBase.java : For the test cases that call the scim2/Me endpoint this base class will be used

3.SCIMProvisioningTestBase.java : This is the base class for the scim test cases that provides the methods for provisioning and deleting users

### CODE REVIEW

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+public class ProcessJsonFile {
+
+    private static String fileName;
+    private static String configFile;
+    private static String rootPath;
+    private static String filePath = "/payload/";
+    static JSONObject rootObject;
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java]: Please remove the additional new line.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package org.wso2.identity.scenarios.commons;
+
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+import java.io.FileReader;
+import java.io.IOException;
+
+public class ProcessJsonFile {
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java]: Shall we add a class comment?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    private static String fileName;
+    private static String configFile;
+    private static String rootPath;
+    private static String filePath = "/payload/";
+    static JSONObject rootObject;
+
+
+    /**
+     *
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java]: Please add the method comment properly and format it.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public static void readFile(String jsonFileName) throws Exception {
+
+        fileName = jsonFileName;
+        rootPath = Thread.currentThread().getContextClassLoader().getResource("").getPath();
+        configFile = rootPath + filePath + fileName;
+    }
+
+    /**
+     *
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java]: Method comment?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java]: Year should be 2019.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package org.wso2.identity.scenarios.commons;
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/ProcessJsonFile.java]: License header?

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+import org.json.simple.JSONObject;
+import org.wso2.identity.scenarios.test.scim2.SCIMConstants;
+import java.io.IOException;
+import static org.wso2.identity.scenarios.commons.util.IdentityScenarioUtil.constructBasicAuthzHeader;
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java]: Remove additional new line.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    public static HttpResponse provisionMeEntity(String serverURL, JSONObject jsonObject, String scimEndPoint, String
+            scimResource, String username, String password) throws IOException {
+
+        HttpClient client = HttpClients.createDefault();
+        scimEndpointURL = serverURL + URL_PATH_SEPARATOR + scimEndPoint + URL_PATH_SEPARATOR + meEndpoint;
+
+        return sendJSONRequest(client, scimEndpointURL, jsonObject, getCommonHeaders(username, password));
+    }
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java]: Remove additional new lines please.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        HttpPost request = new HttpPost(url);
+        if (headers != null) {
+            request.setHeaders(headers);
+        }
+        request.setEntity(new StringEntity(jsonObject.toString()));
+
+        return client.execute(request);
+    }
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java]: Remove additional new lines please.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+      public static HttpResponse provisionMeEntityWithoutAuth(String serverURL, JSONObject jsonObject, String
+                scimEndPoint, String scimResource, String username, String password) throws IOException {
+
+            HttpClient client = HttpClients.createDefault();
+            scimEndpointURL = serverURL + URL_PATH_SEPARATOR + scimEndPoint + URL_PATH_SEPARATOR + meEndpoint;
+
+            return sendJSONRequest(client, scimEndpointURL, jsonObject, getContentTypeHeaders());
+     }
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java]: Remove additional new lines please.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+     */
+    public static Header[] getContentTypeHeaders() {
+
+        Header[] headers = {
+                new BasicHeader(HttpHeaders.CONTENT_TYPE, SCIMConstants.CONTENT_TYPE_APPLICATION_JSON),
+        };
+        return headers;
+
+    }
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIM2TestBase.java]: Remove additional new lines please.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMConstants.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+   public static final String META_CREATED_ATTRIBUTE = "created";
+   public static final String META_CREATED_ATTRIBUTE_VALUE = "2010-01-23T04:56:22Z";
+   public static final String META_RESOURCE_TYPE_ATTRIBUTE = "resourceType";
+   public static final String META_RESOURCE_TYPE_ATTRIBUTE_VALUE = "User";
+   public static final String META_LASTMODIFIED_ATTRIBUTE = "lastModified";
+   public static final String META_LASTMODIFIED_ATTRIBUTE_VALUE = "2011-05-13T04:42:34Z";
+   public static final String META_VERTSION_ATTRIBUTE = "version";
+   public static final String META_VERTSION_ATTRIBUTE_VALUE = "W\\//\"3694e05e9dff591\"";
+   public static final String META_LOCATION_ATTRIBUTE = "location";
+   public static final String META_LOCATION_ATTRIBUTE_VALUE = "https://example.com/v1/Users/2819c223-7f76-453a-919d-413861904646";
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMConstants.java]: We should only have 120 characters per line.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java]: 2019

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    public static Header[] getCommonHeaders(String username, String password) {
+
+        Header[] headers = {
+                new BasicHeader(HttpHeaders.CONTENT_TYPE, SCIMConstants.CONTENT_TYPE_APPLICATION_JSON),
+                new BasicHeader(HttpHeaders.AUTHORIZATION, constructBasicAuthzHeader(username, password))
+        };
+        return headers;
+    }
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java]: Remove additional new lines please.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    public static HttpResponse deleteSCIMEntity(String serverURL, String userID, String scimEndPoint, String scimResource,
+                                                String username, String password) throws IOException {
+
+        HttpClient client = HttpClients.createDefault();
+        scimEndpointURL = serverURL + URL_PATH_SEPARATOR + scimEndPoint + URL_PATH_SEPARATOR + scimResource + URL_PATH_SEPARATOR + userID;
+
+        return sendDeleteRequest(client, scimEndpointURL, getCommonHeaders(username, password));
+    }
+
+
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java]: Remove additional new lines please.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+     * @param serverURL        the server url
+     * @param scimEndPoint     the scim endpoint 1.1 or 2.0
+     * @param scimResource     the scim endpoint for User or Bulk
+     * @param jsonObject       the json object of the scim payload
+     * @param username         Authenticating username.
+     * @param password         Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during user creation.
+     */
+    public static HttpResponse provisionSCIMEntity(String serverURL, JSONObject jsonObject, String scimEndPoint, String
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java]: We should only have 120 characters per line.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+     * @param serverURL        the server url
+     * @param userID           the ID of the user to be deleted
+     * @param username         Authenticating username.
+     * @param scimEndPoint     the scim endpoint for 1.1 or 2.0
+     * @param scimResource the scim endpoint for User or Bulk
+     * @param password         Authenticating user password.
+     * @return HttpResponse with the result.
+     * @throws IOException If error occurs during user deletion.
+     */
+    public static HttpResponse deleteSCIMEntity(String serverURL, String userID, String scimEndPoint, String scimResource,
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMProvisioningTestBase.java]: We should only have 120 characters per line.

File: product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMConstants.java [hunk 18]
```diff
@@ -0,0 +1,159 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (vihanga-liyanage) [inline product-scenarios/scenarios-commons/src/main/java/org/wso2/identity/scenarios/commons/SCIMConstants.java]: 2019

**Comment by vihanga-liyanage** [comment]
@shavantha Can you please address the comments added? We have to get these emerged as soon as possible. :)

**Comment by thanujalk** [comment]
Closed as this already implemented.

**Pull request closed without merging**